### ISSUE TITLE
Tethering Still Shows

### ISSUE THREAD

**Issue by jjbear11** [nneonneo/iOS-SOCKS-Server#18]
Hello! I have been using this program for the last three months without issue. Now, when I try to use it, my data usage shows as Hotspot on my plan (and thus is throttled once I've gone over my cap). I'm on iOS 16.5.1 (c) and recently downloaded the most up-to-date code from the repo to see if any recent changes would fix my issue. Is there something in that version of iOS which could be causing issues here? Thanks for a great little script!

**Comment by jjbear11** [comment]
I am getting the same public IP address whether going through the 'normal' tethering or the iOS-SOCKS-Server script. I did notice however that when I do the 'normal' tethering route, my public IP is not the same on my connected device as it is on the host device. I would think that the aspect of tethering would inherently imply that the public host IP would be shared with the client (ie, I would see the same public IP on my client as I'm seeing on the host).

**Comment by philrosenthal** [comment]
The phone essentially has two cellular interfaces, one for on-device usage, and one for tethering usage. Unfortunately, there's not a clear way of selecting which one is the cellular interface, so we have to make an informed guess.

Each of those interfaces may have both an IPv4 and IPv6 address, or just an IPv4 or just an IPv6 address.

If you are trying to have it bill as if it is usage from the phone, it needs to have the same ipv4 / ipv6 address as the normal on-device usage on the phone.

There were some major changes in the past few months in order to add IPv6 support, and it would be helpful if you could try it at these particular commits to see if any of them pick the 'right' ip address.

https://github.com/nneonneo/iOS-SOCKS-Server/tree/330c97dd050ba5de142b17695407bcceadf32018
https://github.com/nneonneo/iOS-SOCKS-Server/tree/653bdc54b27c91711917657cb8376a4eba06ffae
https://github.com/nneonneo/iOS-SOCKS-Server/tree/46a8ec61b7bce66c5e206b6294a2581f8acb5822
https://github.com/nneonneo/iOS-SOCKS-Server/tree/47f7ce1ef3a14d4deabe053d08f5816943646c79

**Comment by jjbear11** [comment]
I used the following [commit](https://github.com/nneonneo/iOS-SOCKS-Server/tree/653bdc54b27c91711917657cb8376a4eba06ffae) since it was the most-recent one. It appears that IPv6 support isn't working for me. I am receiving an error: **Failed to connect to 2606:4700:4700::1111 over IPv6 due to: [Errno 65] No route to host**. Any suggestions on what I can try?

**Comment by philrosenthal** [comment]
The failure of IPv6 isn't necessarily a problem. If your Cellular carrier doesn't have working IPv6 but does have an IPv6 address configured for some reason, this test takes place to make sure not to attempt to use IPv6 for the rest of the session.

As long as you aren't seeing an IPv6 address show up when you load ip6only.me in a browser, it's safe to ignore this. If you are seeing it show up in your tethering usage, though, the bigger question is figuring out why the IPv4 address is not the IPv4 address that your phone normally uses, but rather the tethering interface's address.

**Comment by jjbear11** [comment]
After some digging, it appears that my mobile provide (AT&T) -- with regards to their hotspot, incorporates double-NAT, or CGNAT technology. Not sure how to work around that... My work mobile is VZW, and I get the same public address on both host hotspot as well as client connected to hotspot. AT&T seems to be the culprit here.

**Comment by nneonneo** [comment]
One option would be to try disabling tethering and use an ad-hoc network instead: create an ad-hoc network on your computer and have your phone join it. iOS-SOCKS-Server has been tested to work in this configuration.

**Comment by philrosenthal** [comment]
> After some digging, it appears that my mobile provide (AT&T) -- with regards to their hotspot, incorporates double-NAT, or CGNAT technology. Not sure how to work around that... My work mobile is VZW, and I get the same public address on both host hotspot as well as client connected to hotspot. AT&T seems to be the culprit here.

CGNAT is the normal way for IPv4 on cellular.

**Comment by philrosenthal** [comment]
@jjbear11 I've just re-read your comment again ... are you saying that the IPv4 address shown on ip4.me is the same on your computer using a hotspot as it is when you run it on your phone in Safari on AT&T?

**Comment by jjbear11** [comment]
So I think I may have stumbled upon something regarding my issue this morning; if I recall correctly, the python script had prior picked **pdp_ip0** as the IPv4 interface. However, now the script picks something besides pdp_ip0 for the IPv4 interface. That appears to cause my tethered traffic to not appear correctly. Is there a way to "force" pdp_ip0 (at least for me to test to see if this is indeed the issue)? Thanks!

**Comment by philrosenthal** [comment]
Try this out and see if it helps.

Look for this section of code (lines 94-100):
        # XXX implement better classification of interfaces
        if iface.name.startswith("en"):
            iftypes["en"].append(iface)
        elif iface.name.startswith("bridge"):
            iftypes["bridge"].append(iface)
        else:
            iftypes["cell"].append(iface)

Change it as follows:

        # XXX implement better classification of interfaces
        if iface.name.startswith("en"):
            iftypes["en"].append(iface)
        elif iface.name.startswith("bridge"):
            iftypes["bridge"].append(iface)
        elif iface.name == 'pdp_ip0':
            iftypes["cell"].append(iface)

**Comment by philrosenthal** [comment]
Can you confirm that this solved the issue with data being tracked as tethering rather than phone usage?

**Comment by jjbear11** [comment]
> Can you confirm that this solved the issue with data being tracked as tethering rather than phone usage?

Yes, I do confirm that hardcoding **pdp_ip0** for the interface name fixed my issue. My data usage properly shows as phone usage now instead of tethering and I am no longer throttled in my speed when using this script. Thanks!

**Comment by philrosenthal** [comment]
@nneonneo I'm not sure your thoughts on making changes here -- this is related to my own experience (and was in my pull request), and as jjbear has noted, the current code is in fact broken for some users (apparently, AT&T), and forcing it to pdp_ip0 does resolve it in that case.

Perhaps there's a way we can look at the routing table to get an idea of what the default route is ... but barring that, I'd be inclined to just hard set it to pdp_ip0 for myself given the clear example shown with this reported issue.

**Comment by nneonneo** [comment]
I'm fine with, say, prioritizing pdp_ip0 over all other options if present. I just don't know if it's possible to guarantee that that will always be the cell connection, but I agree that the current implementation is deficient.

**Comment by philrosenthal** [comment]
I'll see if I can find some good solutions to analyze the routing table, and if not, will select pdp_ip0 as long as it is present and if not, go back to the current algorithm.

Once I've got that completed, I will post a pull request.

**Comment by philrosenthal** [comment]
It looks like there's no straightforward way of accessing the routing table under Pythonista ... 

Perhaps this code could be used as an example on building something to access system libraries directly to see the routing table under Pythonista:

https://github.com/jianpx/ios-cabin/tree/master/RouteTableManager

I'm not sure how difficult it would be to port if it's even possible.

**Comment by nneonneo** [comment]
@philrosenthal I can confirm that it's possible to obtain the route table on iOS, with this proof of concept:

```python
import ctypes

CTL_NET = 4
PF_ROUTE = 17
NET_RT_DUMP2 = 7

mib = (ctypes.c_int * 6)()
mib[0] = CTL_NET
mib[1] = PF_ROUTE
mib[2] = 0  # protocol number, currently always 0
mib[3] = 0  # address family, 0 for all address families
mib[4] = NET_RT_DUMP2
mib[5] = 0

libc = ctypes.CDLL("libSystem.dylib")
sz = ctypes.c_int()
res = libc.sysctl(mib, 6, 0, ctypes.byref(sz), 0, 0)
print(res, sz)

buf = ctypes.create_string_buffer(sz.value)
res = libc.sysctl(mib, 6, buf, ctypes.byref(sz), 0, 0)
print(res, buf.raw, sz)
```

What we need to do now is parse the `buf` according to the binary format into a useful form.
